openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -0,0 +1,147 @@
+"""
+Get call stacks of Model Class
+in three cases-
+1. QuerySet API
+2. save()
+3. delete()
+
+classes:
+CallStackManager -  stores all stacks in global dictionary and logs
+CallStackMixin - used for Model save(), and delete() method
+
+Functions:
+capture_call_stack - global function used to store call stack
+
+Decorators:
+donottrack - mainly for the places where we know the calls. This decorator will let us not to track in specified cases
+
+How to use-
+1. Import following in the file where class to be tracked resides
+    from openedx.core.djangoapps.call_stack_manager import CallStackManager, CallStackMixin
+2. Override objects of default manager by writing following in any model class which you want to track-
+    objects = CallStackManager()
+3. For tracking Save and Delete events-
+    Use mixin called "CallStackMixin"
+    For ex.
+        class StudentModule(CallStackMixin, models.Model):
+4. Decorator is a parameterized decorator with class name/s as argument
+    How to use -
+    1. Import following
+        import from openedx.core.djangoapps.call_stack_manager import donottrack
+"""
+
+import logging
+import traceback
+import re
+import collections
+from django.db.models import Manager
+
+log = logging.getLogger(__name__)
+
+# Module Level variables
+# dictionary which stores call stacks.
+# { "ModelClasses" : [ListOfFrames]}
+# Frames - ('FilePath','LineNumber','Context')
+# ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+#                                                     [(file,line number,context),(---,---,---)]]}
+STACK_BOOK = collections.defaultdict(list)
+
+# filter to trickle down call stacks
+EXCLUDE_REGEX = ['^.*python2.7.*$', '^.*<exec_function>.*$', '^.*exec_code_object.*$', '^.*edxapp/src.*$',
+                 '^.*call_stack_manager(?!/tests).*$']
```

Comment:
Another option here, rather than encoding the test exclusion here, would be to override `EXCLUDE_REGEX` and `REGULAR_EXPS` in your tests so that none of the `call_stack_manager` lines get excluded while the tests are running.